teamcity: failed tests on master: test/TestNodeIsLiveCallback

The following tests appear to have failed:

[#381431](https://teamcity.cockroachdb.com/viewLog.html?buildId=381431):

```
--- FAIL: test/TestNodeIsLiveCallback (45.170s)
soon.go:44: storage/node_liveness_test.go:190, condition failed to evaluate within 45s: expected IsLive callback for node 1
	goroutine 96875 [running]:
	runtime/debug.Stack(0x2, 0x2925577, 0x1d)
		/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
	github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoonDepth(0x1, 0x2c192c0, 0xc4230f6000, 0xc423303f78)
		/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:45 +0xa1
	github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2c192c0, 0xc4230f6000, 0xc423303f78)
		/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x48
	github.com/cockroachdb/cockroach/pkg/storage_test.TestNodeIsLiveCallback(0xc4230f6000)
		/go/src/github.com/cockroachdb/cockroach/pkg/storage/node_liveness_test.go:190 +0x3c2
	testing.tRunner(0xc4230f6000, 0x1e3fe10)
		/usr/local/go/src/testing/testing.go:746 +0xd0
	created by testing.(*T).Run
		/usr/local/go/src/testing/testing.go:789 +0x2de
------- Stdout: -------
I171018 16:26:14.327307 96875 storage/store.go:1212  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I171018 16:26:14.327359 96875 gossip/gossip.go:331  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:32928" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
W171018 16:26:14.330483 96875 gossip/gossip.go:1277  [n2] no incoming or outgoing connections
I171018 16:26:14.330688 97608 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:32928
I171018 16:26:14.331186 96875 gossip/gossip.go:331  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:42793" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
W171018 16:26:14.334072 96875 gossip/gossip.go:1277  [n3] no incoming or outgoing connections
I171018 16:26:14.334169 96875 storage/store.go:1212  [s3] [n3,s3]: failed initial metrics computation: [n3,s3]: system config not yet available
I171018 16:26:14.334207 96875 gossip/gossip.go:331  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:37719" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I171018 16:26:14.338647 97828 gossip/client.go:129  [n3] started gossip client to 127.0.0.1:32928
I171018 16:26:36.923080 97249 gossip/gossip.go:1291  [n2] node has connected to cluster via gossip
I171018 16:26:38.151013 97626 gossip/gossip.go:1291  [n1] node has connected to cluster via gossip
I171018 16:26:42.559997 97850 gossip/gossip.go:1291  [n3] node has connected to cluster via gossip
I171018 16:26:59.497607 97611 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:37719->127.0.0.1:56907: use of closed network connection
I171018 16:26:59.497615 97714 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:42793->127.0.0.1:59951: use of closed network connection
I171018 16:26:59.497658 97515 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:32928->127.0.0.1:40395: use of closed network connection
I171018 16:26:59.497690 97619 vendor/google.golang.org/grpc/grpclog/grpclog.go:60  transport: http2Client.notifyError got notified that the client transport was broken EOF.
W171018 16:26:59.497732 97850 gossip/gossip.go:1277  [n3] no incoming or outgoing connections

```
Please assign, take a look and update the issue accordingly.

--- a/pkg/storage/node_liveness.go
+++ b/pkg/storage/node_liveness.go
@@ -692,13 +692,30 @@ func (nl *NodeLiveness) updateLivenessAttempt(
 	return nil
 }
 
+// maybeUpdate replaces the liveness (if it appears newer) and invokes the
+// registered callbacks if the node became live in the process.
 func (nl *NodeLiveness) maybeUpdate(new Liveness) {
 	nl.mu.Lock()
+	// Note that this works fine even if `old` is empty.
 	old := nl.mu.nodes[new.NodeID]
-	if shouldReplaceLiveness(old, new) {
+	should := shouldReplaceLiveness(old, new)
+	var callbacks []IsLiveCallback
+	if should {
 		nl.mu.nodes[new.NodeID] = new
+		callbacks = append(callbacks, nl.mu.callbacks...)
 	}
 	nl.mu.Unlock()
+
+	if !should {
+		return
+	}
+
+	now, offset := nl.clock.Now(), nl.clock.MaxOffset()
+	if !old.IsLive(now, offset) && new.IsLive(now, offset) {
+		for _, fn := range callbacks {
+			fn(new.NodeID)
+		}
+	}
 }
 
 func shouldReplaceLiveness(old, new Liveness) bool {
@@ -734,28 +751,7 @@ func (nl *NodeLiveness) livenessGossipUpdate(key string, content roachpb.Value) 
 		return
 	}
 
-	// If there's an existing liveness record, only update the received
-	// timestamp if this is our first receipt of this node's liveness, the
-	// expiration or epoch was advanced, or the draining state changed.
-	var callbacks []IsLiveCallback
-	nl.mu.Lock()
-	exLiveness := nl.mu.nodes[liveness.NodeID]
-	if shouldReplaceLiveness(exLiveness, liveness) {
-		nl.mu.nodes[liveness.NodeID] = liveness
-
-		// If isLive status is now true, but previously false, invoke any registered callbacks.
-		// Note that this works fine even if exLiveness is empty.
-		now, offset := nl.clock.Now(), nl.clock.MaxOffset()
-		if !exLiveness.IsLive(now, offset) && liveness.IsLive(now, offset) {
-			callbacks = append(callbacks, nl.mu.callbacks...)
-		}
-	}
-	nl.mu.Unlock()
-
-	// Invoke any "is live" callbacks after releasing lock.
-	for _, cb := range callbacks {
-		cb(liveness.NodeID)
-	}
+	nl.maybeUpdate(liveness)
 }
 
 // numLiveNodes is used to populate a metric that tracks the number of live

